Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [1]:
NAME = "Maryana Shnitser"
COLLABORATORS = "Adrianna Celtruda"

---

# Assignment: game2

In this assignment we will be expanding on our original Zork game. 

**Few Important Things:**
 - Save this file as **geo3_emailusername_assignment (e.g. geo3_mmann1123_game2.ipynb)**
 - Before submitting make sure to go to Cell > Run All and confirm there are no errors thrown.
 - If you can't figure something out, that is ok, do your best, feel free to post a specific question on the blackboard Q&A discuession board. 


**How these assignments will work: For detailed instructions on how to complete these exercises please read the following [tutorial](https://mmann1123.github.io/GEOG6308/)**. 

## More Text Games

For this exercise we are going to be building a new [Zork-like](https://textadventures.co.uk/games/play/5zyoqrsugeopel3ffhz_vq) game.  

To start, we are going to build some game components!

----
## Part (5 points)

What's a game without enemies?!  Most text based games have enemies that have a few properties. We are going to keep things very simple. Since we haven't learned about `classes` yet, we are going to use dictionaries to keep track of ourselves and enemy characteristics. 

Define a function called `new_ogre` to create a dictionary with the character properities. 
The function should have:
- one parameter called `xp_value` that takes any number, 
- with a default value of 50. 
- the function should `return` a dictionary called `ogre`
    - The dictionary, called for instance `ogre`, stores two characteristics as key value pairs:
        - `health` points set initially to 50
        - `name` set to "ogre"

HINT: (See section called "Default Values" in textbook)


In [2]:
"""Define a function called new_ogre that returns a dictionary ogre"""

# YOUR CODE HERE
def new_ogre (xp_value=50):
    ogre = {'health': xp_value, 'name': 'ogre'}
    return ogre


Use the function `new_self` to create a dictionary called that stores two characteristics:
- `health` points are always set to 100 (no parameter needed for this one)
- `name` set to "self"


In [3]:
"""Define a function called new_self that returns a dictionary myself"""

# YOUR CODE HERE
def new_self():
    myself ={'health':100, 'name' : 'self'}
    return myself

Now we are going to create two characters `enemy1` who is an ogre and `myself` who is... well you...

In [4]:
'''Run your functions and store output'''

enemy1 = new_ogre()
myself = new_self()


In [5]:
"""Check that new_characters returns the correct output"""
try:
    assert isinstance(new_ogre(),dict) 
except AssertionError:
    raise AssertionError("Your enemy is not a dictionary, please make it so")
try:
    assert isinstance(new_self(),dict) 
except AssertionError:
    raise AssertionError("myself is not a dictionary, please make it so")     

In [6]:
try:
    assert new_ogre()['health'] == 50 
except AssertionError:
    raise AssertionError("incorrect default value for your ogre xp_value")  
try:
    assert new_ogre(25)['health'] == 25 
except AssertionError:
    raise AssertionError("unable to pass xp_value as an argument to new_ogre")   
try:
    assert new_ogre()['name'] == 'ogre' 
except AssertionError:
    raise AssertionError("new_ogre name should always be set to 'ogre'")  

In [7]:
try:
    assert new_self()['health'] == 100 
except AssertionError:
    raise AssertionError("new_self health should always be set to 100")  
try:
    assert new_self()['name'] == 'self' 
except AssertionError:
    raise AssertionError("new_self name should always be set to 'self'")  

In [8]:
try:
    assert [i for i in enemy1.keys()] == ['health', 'name'] or [i for i in enemy1.keys()] == ['name', 'health']
except AssertionError:
    raise AssertionError("enemy doesn't contain the right keys")
try:
    assert [i for i in myself.keys()] == ['health', 'name'] or [i for i in myself.keys()] == ['name', 'health']
except AssertionError:
    raise AssertionError("enemy doesn't contain the right keys")

In [9]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [10]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [11]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

----
## Part (5 points)

Ok now we need to be able to attack!  Let's create a function called `attack` that has two parameters:
- one called `who` which will accept the dictionary defining the person being attacked
- and a second called `strength` that takes a numeric value indicating how many xp_points will be lost
    - the default `strength` should be 20
    
When executed the `attack` function should:
- Subtract the value of `strength` from the value stored under the character's dictionary called `health`
    - This new value should then be stored under `health` in the dictionary
- The function should return back the updated `who` dictionary

For instance, assuming enemy1 health is 100, `attack(who=enemy1,strength=10)` should return the following dictionary `{'name':ogre,'health':90}` *the first time you run it*

In [12]:
"""Define a function called new_ogre that returns a dictionary ogre"""

def attack(who, strength = 20):
   who['health'] -= strength
   if who['health'] <= 0:
      who['health'] = 0  # Set health to 0 if it falls below or equal to 0
   return who

In [13]:
# attack the enemy! This should subtract 50 health points from enemy1's current health value.
enemy1 = attack(who=enemy1, strength=50)
enemy1


{'health': 0, 'name': 'ogre'}

In [14]:
"""Check that attack returns the correct output"""
try:
    assert attack(new_ogre(xp_value=100), strength=50)['health'] == 50
except AssertionError:
    raise AssertionError("You aren't subtracting the right amount of health, or new_orgre() isn't working")

In [15]:
try:
    assert attack(new_ogre(xp_value=50), strength=50)['health'] == 0
except AssertionError:
    raise AssertionError("You aren't subtracting the right amount of health, or new_orgre() isn't working")

In [16]:
try:
    assert attack(new_ogre(xp_value=100))['health'] == 80
except AssertionError:
    raise AssertionError("You aren't subtracting the right amount of health, or new_orgre() isn't working")


In [17]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [18]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [19]:
try:
    assert attack(new_ogre(xp_value=50), strength=100)['health'] == 0
except AssertionError:
    raise AssertionError("You aren't subtracting the right amount of health, or new_orgre() isn't working")

----
## Part (6 points)

Ok but.... you've programmed the undead here.... really man. Negative health is a problem that should be fixed. Let's update our attack function so that we can kill our enemies. 

Add an `if elif else` statement to your `attack` function. Make sure to do the following things:

- If the enemy (defined by `who`) already has a `health` <= 0
    - print out the following `"What is wrong with you man... No hitting dead bodies!"`
    - return who, but make sure `health = 0`
- else if `who`s `health` is <= `strength`:
    - print "You've killed the (insert the correct `name` here)!"
    - set `health` to 0
    - return `who`
- else
    - print "Good hit, the (insert `name`) still has (insert `health` value) health points left."
    - update the `health` value to reflect the `strength` of the attack
    - return `who`
    
  
     

In [20]:
"""Define update your attack function to handle death"""

# YOUR CODE HERE
def attack(who, strength = 10):
    if who['health']<=0:
        print ("What is wrong with you man... No hitting dead bodies!")
        who['health'] = 0 
        return who
    elif who['health']<= strength:
        print (f"Youve killed the {who}!")
        who ['health'] = 0 
        return who
    else: 
        print (f"Good hit, the {who} still has {who['health']} health points left")
        who['health'] -= strength
        return who

In [21]:
# create a new ogre with full health
enemy2 = new_ogre(xp_value=50)

# attack the enemy
enemy2 = attack(enemy2, strength=20)

# show result (health should be 30)
enemy2

Good hit, the {'health': 50, 'name': 'ogre'} still has 50 health points left


{'health': 30, 'name': 'ogre'}

In [22]:
"""Check that new attack returns the correct output"""
try:
    assert attack(new_ogre(xp_value=100), strength=50)['health'] == 50
except AssertionError:
    raise AssertionError("You aren't subtracting the right amount of health, or new_orgre() isn't working")

Good hit, the {'health': 100, 'name': 'ogre'} still has 100 health points left


In [23]:

import inspect
# custom function to check answers
def check_if_elif_else_used(func, check_if=True, check_if_elif=False, check_if_elif_else=False):
    """
    Check if the function contains if, elif, and else statements,
    ignoring any lines that are commented out.
    """
    # Get the source code of the function
    source = inspect.getsource(func)
    
    # Split the source into lines and remove comments
    uncommented_lines = []
    for line in source.splitlines():
        # Strip leading and trailing whitespaces
        stripped_line = line.strip()
        
        # Ignore lines that start with a comment
        if not stripped_line.startswith("#"):
            uncommented_lines.append(stripped_line)
    
    # Join uncommented lines into a single text
    uncommented_source = " ".join(uncommented_lines)
    
    # Check for the presence of "if", "elif", and "else"
    has_if = "if " in uncommented_source
    has_elif = "elif " in uncommented_source
    has_else = "else:" in uncommented_source
    
    if check_if_elif_else:
        return has_if and has_elif and has_else
    elif check_if_elif:
        return has_if and has_elif
    elif check_if:
        return has_if
    else:
        return None

assert check_if_elif_else_used(attack, check_if_elif_else=True) == True, "Test failed! The function does not contain both if, elif, else statements."


In [24]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [25]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [26]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [27]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [28]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

---
## Part (10 points)

Set up a single room called `enemyroom` that:
- Accepts two arguments via `yourself` and `enemy` e.g. `def enemyroom(yourself, enemy)``
- You encounter an enemy, remember to create a new one (e.g. anenemy = new_ogre(xp_value=50))
- User is given the chance to:
    - Run away or
        - You live to fight another day. 
    - Fight
        - If they decide to fight, you and the enemy should engage in mortal combat. 
        - Two people enter, one leaves.
- Return `winner` dictionary from the room - even if you just ran away like a coward.
- Print some message based on `who` is returned. 
    - Congraduate the winner!

HINTS: 
- For prompts please use capital "Y" and "N" for all user interactions e.g. "Do you want to attack (Y or N)?"
- Use a while loop for the fight sequence that stops only if someone dies. 
- Use the `attack` function you created before to manage the fight sequence
- Typically attacks go in rounds, where you have a chance to strike, and the opponent has a chance to attack -  in each round of the loop.
- You can create a random `strength` of attack using `random.randrange(0,50,1)`, which creates a random integer between 0 and 50, the `1` is the step (e.g. If you choose a step of 15 `random.randrange(0,50,15)` it can only generate the numbers `[0,15,30,45]`).  JUST MAKE SURE TO `import random` at the top of your cell. 
- Make sure you can kill the enemy or the enemy can kill you.


In [29]:
"""Write your asnwer here. """
import random

# hint 
# pass yourslef and the enemy into the room
# def enemyroom(yourself, enemy):
#     print('You encounter an ugly ass ogre... or something...')
    
# YOUR CODE HERE
winner = {'name', 'result'}
yourself = {'health':(random.randrange(0,50,15)), 'name':'yourself'}
enemy = {'health':(random.randrange(0,50,15)), 'name':'enemy'} 

def enemyroom(yourself, enemy): 

    message = input('You encounter an ugly ass ogre. Do you want to fight (Y or N)') 
    

    if (message == 'N'):
        print ('You live to fight another day')
        return {'name': yourself['name'], 'result':'Winner', 'health':yourself['health']}
    
    elif (message == 'Y'):
        while yourself['health']>0 and enemy['health']>0:

            enemy = attack(who= enemy, strength = random.randrange(0,50,1))
            if enemy['health'] == 0:
                print(f"You killed the {enemy['name']}")
                print(f"{yourself['name']} won the battle") 
                return {'name': yourself['name'], 'result': 'Winner', 'health': yourself['health']}
        
            yourself = attack (who= yourself, strength = random.randrange(0,50,1))
            if yourself['health'] == 0:
                print (f"{yourself['name']} died")
                print(f"The winner is {enemy['name']}!")
                return {'name': enemy['name'], 'result': 'Winner', 'health':yourself['health']}
        return winner
        print (winner)


In [30]:
# Example
# create new characters
anenemy = new_ogre(xp_value=50)
me = new_self()

# enter room
# enemyroom(yourself = me, enemy = anenemy)   # please comment this out before submiting for grading!

**Grading Criteria (10.5 points)**

To grade this section I am using the `unittest` module to simulate user interactions with you game. 

For instance the following will simulate me typing "Y" after running the room function:

``` python
with patch('builtins.input', side_effect=['Y']):
```

- [ ] Check that someone's dictionary is returned from the room  (eg someone walks away)try:
    assert isinstance(new_ogre(),dict) 
except AssertionError:
    raise AssertionError("Your enemy is not a dictionary, please make it so")

- [ ] User prompted to fight or run - with valid responses to both 

- [ ] An extremely strong character should always win 

- [ ] The winner should have 0 or more health

In [31]:
# The following defines tests used for grading

# if importing unittest throws an error, open a new terminal from vscode or jupyter lab, make sure you are in the right environment
# and run `pip install unittest2`
from unittest.mock import patch
import unittest



class TestEnemyRoom(unittest.TestCase): 
    def test_there_is_a_winner(self):
        player = {'name': 'player', 'health': 100}
        enemy = {'name': 'ogre', 'health': 100}
        winner = enemyroom(player, enemy)
        self.assertIsInstance(winner,dict, "There much be a winner, please return a characters dictionary.")

    def test_encounter_weak_enemy(self):
        player = {'name': 'player', 'health': 1000}
        enemy = {'name': 'ogre', 'health': 10}
        winner = enemyroom(player, enemy)
        print('winner is', winner)
        # check that the player wins
        self.assertTrue(winner['name'] == 'player', "The player should win since orgre was SO weak. Check your code!")
     
    def test_encounter_strong_enemy(self):
        player = {'name': 'player', 'health': 10}
        enemy = {'name': 'ogre', 'health': 1000}
        winner = enemyroom(player, enemy)
        print('winner is', winner)
        self.assertTrue(winner['name'] == 'ogre', "The ogre should win the player was SO weak. Check your code!")
        
    def test_winner_isnt_undead(self):
        player = {'name': 'player', 'health': 200}
        enemy = {'name': 'ogre', 'health': 100}
        winner = enemyroom(player, enemy)
        print('winner is', winner)
        print(f'The winner has {winner["health"]} health left.')
        self.assertTrue(winner['health'] >= 0, "The winner should have 0 or more health left.")

    def test_winner_returned(self):
        player = {'name': 'player', 'health': 100}
        enemy = {'name': 'ogre', 'health': 100}
        winner = enemyroom(player, enemy)
        self.assertIn('name', winner, "The winner should be returned and have a name.")

  
# Manually create an instance of TestPickUpItem
test_case_instance = TestEnemyRoom()


In [32]:
""" Tests: do not edit or remove """
try:
    assert isinstance(new_ogre(),dict) 
except AssertionError:
    raise AssertionError("Your enemy is not a dictionary, please make it so")

In [33]:
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['Y']):
    print('#####################\n Testing that a winner is returned after a fight \n #####################.')
    test_case_instance.test_there_is_a_winner()

#####################
 Testing that a winner is returned after a fight 
 #####################.
Good hit, the {'name': 'ogre', 'health': 100} still has 100 health points left
Good hit, the {'name': 'player', 'health': 100} still has 100 health points left
Good hit, the {'name': 'ogre', 'health': 70} still has 70 health points left
Good hit, the {'name': 'player', 'health': 68} still has 68 health points left
Good hit, the {'name': 'ogre', 'health': 63} still has 63 health points left
Good hit, the {'name': 'player', 'health': 48} still has 48 health points left
Good hit, the {'name': 'ogre', 'health': 46} still has 46 health points left
Good hit, the {'name': 'player', 'health': 46} still has 46 health points left
Good hit, the {'name': 'ogre', 'health': 29} still has 29 health points left
Good hit, the {'name': 'player', 'health': 34} still has 34 health points left
Youve killed the {'name': 'ogre', 'health': 8}!
You killed the ogre
player won the battle


In [34]:

# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['N']):
    print('#####################\n Testing that a coward is returned after he runs away \n #####################.')
    test_case_instance.test_there_is_a_winner()

#####################
 Testing that a coward is returned after he runs away 
 #####################.
You live to fight another day


In [35]:
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['Y']):
    print('#####################\n Testing that the player wins if they encounter a weak enemy \n #####################.')
    test_case_instance.test_encounter_weak_enemy()

#####################
 Testing that the player wins if they encounter a weak enemy 
 #####################.
Youve killed the {'name': 'ogre', 'health': 10}!
You killed the ogre
player won the battle
winner is {'name': 'player', 'result': 'Winner', 'health': 1000}


In [36]:
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['Y']):
    print('#####################\n Testing that the player wins if they encounter a strong enemy \n #####################.')
    test_case_instance.test_encounter_strong_enemy()

#####################
 Testing that the player wins if they encounter a strong enemy 
 #####################.
Good hit, the {'name': 'ogre', 'health': 1000} still has 1000 health points left
Youve killed the {'name': 'player', 'health': 10}!
player died
The winner is ogre!
winner is {'name': 'ogre', 'result': 'Winner', 'health': 0}


In [37]:
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['Y']):
    print('#####################\n Testing that the player wins if they encounter a strong enemy \n #####################.')
    test_case_instance.test_winner_isnt_undead()

#####################
 Testing that the player wins if they encounter a strong enemy 
 #####################.
Good hit, the {'name': 'ogre', 'health': 100} still has 100 health points left
Good hit, the {'name': 'player', 'health': 200} still has 200 health points left
Good hit, the {'name': 'ogre', 'health': 88} still has 88 health points left
Good hit, the {'name': 'player', 'health': 197} still has 197 health points left
Good hit, the {'name': 'ogre', 'health': 58} still has 58 health points left
Good hit, the {'name': 'player', 'health': 157} still has 157 health points left
Youve killed the {'name': 'ogre', 'health': 10}!
You killed the ogre
player won the battle
winner is {'name': 'player', 'result': 'Winner', 'health': 118}
The winner has 118 health left.


In [38]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [39]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [40]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [41]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['Y']):
    print('#####################\n Testing that the player wins if they encounter a strong enemy \n #####################.')
    test_case_instance.test_encounter_strong_enemy()

#####################
 Testing that the player wins if they encounter a strong enemy 
 #####################.
Good hit, the {'name': 'ogre', 'health': 1000} still has 1000 health points left
Youve killed the {'name': 'player', 'health': 10}!
player died
The winner is ogre!
winner is {'name': 'ogre', 'result': 'Winner', 'health': 0}


In [42]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['Y']):
    print('#####################\n Testing that the player wins if they encounter a strong enemy \n #####################.')
    test_case_instance.test_winner_isnt_undead()

#####################
 Testing that the player wins if they encounter a strong enemy 
 #####################.
Good hit, the {'name': 'ogre', 'health': 100} still has 100 health points left
Good hit, the {'name': 'player', 'health': 200} still has 200 health points left
Good hit, the {'name': 'ogre', 'health': 69} still has 69 health points left
Good hit, the {'name': 'player', 'health': 182} still has 182 health points left
Good hit, the {'name': 'ogre', 'health': 57} still has 57 health points left
Good hit, the {'name': 'player', 'health': 137} still has 137 health points left
Good hit, the {'name': 'ogre', 'health': 30} still has 30 health points left
Good hit, the {'name': 'player', 'health': 88} still has 88 health points left
Youve killed the {'name': 'ogre', 'health': 22}!
You killed the ogre
player won the battle
winner is {'name': 'player', 'result': 'Winner', 'health': 79}
The winner has 79 health left.


In [43]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

---
## Challenge Question (3 points)

A common element of text based games is the ability to pickup items and store them for use later. In this section we are going to set up the ability to pick up items using a `while` loop in the function called `pick_up_item`. 

We are going to use a `dictionary` to act as the backpack for our character. The backpack has the following properties:
- Every item *available to be picked up* will be stored in a dictionary called  `items_in_chest`.
    - The `items_in_chest` dictionary should store the item name and it's weight. 
- We will store the items in our backpack as we pick them up within `items_in_backpack`.
- The backpack can only hold 3 items at any time, to do this we can use the length of `len(items_in_backpack)`, which tells us how many keys are in our dictionary. 
- We need to prompt the user multiple times to see if they want to pick an item up, to do this we will use a `while` loop.


Hint: The `while` loop should run until the backpack is full, therefore the while loop should be based on the len of the backpack dictionary. See section called "The while Loop in Action" pg 118


In [44]:
# a tresure chest with 4 items and their weights
items_in_chest = {'blanket':3, 'short sword':35, 'dagger':5, 'chain mail':12 }

# your currently empty backpack
items_in_backpack = {}

import time
print(f"You wake up dazed.")
time.sleep(1)
print("Ugh... your still drunk... your head is ringing")
print('\a')
time.sleep(2)
print("What's that taste in your mouth?\nYou promptly vomit all over yourself")
time.sleep(2)
print(f"Oh man what's this? You open a large box sitting next you you and see the following items:\n{[i for i in items_in_chest]}")
print('This morning took a sudden turn for the better!')
time.sleep(2)
print("Now if only you could figure out how to pick these items up...")

You wake up dazed.
Ugh... your still drunk... your head is ringing

What's that taste in your mouth?
You promptly vomit all over yourself
Oh man what's this? You open a large box sitting next you you and see the following items:
['blanket', 'short sword', 'dagger', 'chain mail']
This morning took a sudden turn for the better!
Now if only you could figure out how to pick these items up...


In [45]:
"""Write your asnwer here. """

# we need to pass items_in_chest and items_in_backpack to the pick_up_item function in order to use them.
# functions only have access to objects that you hand them.

def pick_up_item(items_in_chest, items_in_backpack): # creates function pick_up_item
    print('In the corner of the room you see the following items in a treasure chest:')

    while len(items_in_backpack) < 3:
        print(f'\nYou see the following items in a treasure chest: {items_in_chest}') # presents user with options of what to pick up
        message = input('What would you like to pick up?\t') #asks for user input

        if (message in items_in_chest) and (message not in items_in_backpack): #if item is in just and not yet in backpack, it can be picked up and added to backpack
            items_in_backpack[message] = items_in_chest[message]
            del items_in_chest[message]
            print(f'You chose {message} and added it to your backpack. The following items are still avaialble to pick up: {items_in_chest}')
        elif message in items_in_backpack: # doesn't allow user to pick up item already in backpack
            print("You've already picked up that item. Please choose a different one.")
        else: #doesn't allow user to pick up items not in the chest
                print(f'The item is not available, try again')

        print (f"You picked up the following items: {items_in_backpack}")
    return items_in_backpack


 

In [1185]:
# Example
items_in_backpack = pick_up_item(items_in_chest, items_in_backpack) # please comment this out before submitting thanks! 

In the corner of the room you see the following items in a treasure chest:

You see the following items in a treasure chest: {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
The item is not available, try again
You picked up the following items: {}

You see the following items in a treasure chest: {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
The item is not available, try again
You picked up the following items: {}

You see the following items in a treasure chest: {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}


### Testing 
The following code tests your `pick_up_items` code. Please confirm that you can pass all the tests 

In [1186]:
from unittest.mock import patch

# Assuming TestPickUpItem and pick_up_item are already defined
class TestPickUpItem(unittest.TestCase):
    # check that returned value is a dictionary
    def test_pick_up_item_returns_dictionary(self):
        print('===========\nTesting that pick_up_item returns a dictionary.\n===========')
        items_in_chest = {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
        items_in_backpack = {}
        
        result = pick_up_item(items_in_chest, items_in_backpack)
        
        self.assertIsInstance(result, dict, "The function should return a dictionary.")


    def test_pick_up_only_three_items(self):
        print('===========\nTesting that you can only pick up 3 items.\n===========')
        items_in_chest = {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
        items_in_backpack = {}
        
        result = pick_up_item(items_in_chest, items_in_backpack)
        
        if len(result) != 3:
            self.fail("You should only be able to pick up 3 items.")
            
        self.assertEqual(result, {'blanket': 3, 'short sword': 35, 'dagger': 5}, "The items in the backpack are incorrect.")

    # try to pick up nonexistent item, then pick up 3 items
    def test_pick_up_nonexistent_item(self):
        print('===========\nTesting that you reprompt user if they try to pick up a nonexistent item.\n===========')
        items_in_chest = {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
        items_in_backpack = {}
        
        result = pick_up_item(items_in_chest, items_in_backpack)
        
        if len(result) != 3:
            self.fail("You should continue to prompt the user if they try to pick up a nonexistent item.")
        
        self.assertEqual(result, {'blanket': 3, 'short sword': 35,'chain mail': 12}, "The items in the backpack are incorrect.")

# Manually create an instance of TestPickUpItem
test_case_instance = TestPickUpItem()

In [1187]:
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['blanket', 'short sword', 'dagger']):
    test_case_instance.test_pick_up_item_returns_dictionary()

Testing that pick_up_item returns a dictionary.
In the corner of the room you see the following items in a treasure chest:

You see the following items in a treasure chest: {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
You chose blanket and added it to your backpack. The following items are still avaialble to pick up: {'short sword': 35, 'dagger': 5, 'chain mail': 12}
You picked up the following items: {'blanket': 3}

You see the following items in a treasure chest: {'short sword': 35, 'dagger': 5, 'chain mail': 12}
You chose short sword and added it to your backpack. The following items are still avaialble to pick up: {'dagger': 5, 'chain mail': 12}
You picked up the following items: {'blanket': 3, 'short sword': 35}

You see the following items in a treasure chest: {'dagger': 5, 'chain mail': 12}
You chose dagger and added it to your backpack. The following items are still avaialble to pick up: {'chain mail': 12}
You picked up the following items: {'blanket': 3, 's

In [1188]:
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['blanket', 'short sword', 'dagger']):
    test_case_instance.test_pick_up_only_three_items()

Testing that you can only pick up 3 items.
In the corner of the room you see the following items in a treasure chest:

You see the following items in a treasure chest: {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
You chose blanket and added it to your backpack. The following items are still avaialble to pick up: {'short sword': 35, 'dagger': 5, 'chain mail': 12}
You picked up the following items: {'blanket': 3}

You see the following items in a treasure chest: {'short sword': 35, 'dagger': 5, 'chain mail': 12}
You chose short sword and added it to your backpack. The following items are still avaialble to pick up: {'dagger': 5, 'chain mail': 12}
You picked up the following items: {'blanket': 3, 'short sword': 35}

You see the following items in a treasure chest: {'dagger': 5, 'chain mail': 12}
You chose dagger and added it to your backpack. The following items are still avaialble to pick up: {'chain mail': 12}
You picked up the following items: {'blanket': 3, 'short 

In [1189]:
# Side effects are passed as if they were user input
with patch('builtins.input', side_effect=['non-existant item','blanket', 'short sword', 'chain mail']):
    test_case_instance.test_pick_up_nonexistent_item()

Testing that you reprompt user if they try to pick up a nonexistent item.
In the corner of the room you see the following items in a treasure chest:

You see the following items in a treasure chest: {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
The item is not available, try again
You picked up the following items: {}

You see the following items in a treasure chest: {'blanket': 3, 'short sword': 35, 'dagger': 5, 'chain mail': 12}
You chose blanket and added it to your backpack. The following items are still avaialble to pick up: {'short sword': 35, 'dagger': 5, 'chain mail': 12}
You picked up the following items: {'blanket': 3}

You see the following items in a treasure chest: {'short sword': 35, 'dagger': 5, 'chain mail': 12}
You chose short sword and added it to your backpack. The following items are still avaialble to pick up: {'dagger': 5, 'chain mail': 12}
You picked up the following items: {'blanket': 3, 'short sword': 35}

You see the following items in a tre

In [1190]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [1191]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '

In [1192]:
""" Hidden tests - These tests mirror the tests above - you will not see any additional feedback """

' Hidden tests - These tests mirror the tests above - you will not see any additional feedback '